addons/mail/controllers/thread.py

```diff
@@ -215,13 +225,41 @@ def mail_message_post(self, thread_model, thread_id, post_data, context=None, **
             raise NotFound()
         if not self._get_thread_with_access(thread_model, thread_id, mode="write"):
             thread = thread.with_context(mail_post_autofollow_author_skip=True, mail_post_autofollow=False)
-        # sudo: mail.thread - users can post on accessible threads
-        message = thread.sudo().message_post(
-            **self._prepare_message_data(post_data, thread=thread, from_create=True, **kwargs),
-        )
+        message_data = self._prepare_message_data(post_data, thread=thread, from_create=True, **kwargs)
+        guest = self.env["mail.guest"]._get_guest_from_context()
+        create_pending = create_pending and not (request.env.user._is_public() and guest)
+        if create_pending:
+            notification_parameters = {}
+            for key in list(message_data.keys()):
+                if key not in self.env["mail.scheduled.message"]._fields:
+                    param = message_data.pop(key)
+                    if (
+                        key
+                        in self.env["mail.scheduled.message"]._notification_parameters_whitelist()
+                    ):
+                        notification_parameters[key] = param
+            message_data["author_id"], email_from = self.env["mail.thread"]._message_compute_author(
+                message_data.get("author_id"), notification_parameters.get("email_from")
+            )
+            # delay by 15 seconds to allow quick cancellation
+            message_data["scheduled_date"] = fields.Datetime.now() + timedelta(
+                seconds=create_pending_post_delay_seconds
+            )
+            if notification_parameters:
+                notification_parameters["email_from"] = email_from
+                message_data["notification_parameters"] = json.dumps(notification_parameters)
+            message = self.env["mail.scheduled.message"].sudo().create({
```

Comment:
I think it's a bad idea to have the same variable hold 2 types of messages.
We should have one key specific for each type.
Then we no longer need the "pending" boolean, as we would have an explicit either message either scheduled_message.